src/main/kotlin/com/expedia/graphql/schema/extensions/annotationExtensions.kt

```diff
@@ -69,32 +69,38 @@ internal fun KAnnotatedElement.isGraphQLID() = this.findAnnotation<GraphQLID>()
 internal fun Annotation.getDirectiveInfo(): DirectiveAnnotation? =
     this.annotationClass.annotations.find { it is DirectiveAnnotation } as? DirectiveAnnotation
 
-internal fun KAnnotatedElement.directives() =
+internal fun KAnnotatedElement.directives(config: SchemaGeneratorConfig) =
     this.annotations.asSequence()
-        .mapNotNull { it.getDirectiveInfo() }
-        .map { it.getGraphQLDirective() }
+        .mapNotNull { it.getGraphQLDirective(config) }
         .toList()
 
 @Throws(CouldNotGetNameOfAnnotationException::class)
-private fun DirectiveAnnotation.getGraphQLDirective(): GraphQLDirective {
-    val kClass: KClass<out DirectiveAnnotation> = this.annotationClass
+private fun Annotation.getGraphQLDirective(config: SchemaGeneratorConfig): GraphQLDirective? {
+    val directiveInfo = getDirectiveInfo() ?: return null
+
+    val kClass = this.annotationClass
+
     val builder = GraphQLDirective.newDirective()
-    val name: String = if (this.name.isNotEmpty()) {
-        this.name
+    val name: String = if (directiveInfo.name.isNotEmpty()) {
+        directiveInfo.name
     } else {
         kClass.simpleName ?: throw CouldNotGetNameOfAnnotationException(kClass)
     }
 
     @Suppress("Detekt.SpreadOperator")
     builder.name(normalizeDirectiveName(name))
-        .validLocations(*this.locations)
-        .description(this.description)
+        .validLocations(*directiveInfo.locations)
+        .description(directiveInfo.description)
+
+    kClass.getValidProperties().forEach { property ->
+        val propertyName = property.name
+        val value = property.call(this)
 
-    kClass.getValidFunctions().forEach { kFunction ->
-        val propertyName = kFunction.name
-        val value = kFunction.call(kClass)
         @Suppress("Detekt.UnsafeCast")
-        val type = defaultGraphQLScalars(kFunction.returnType) as GraphQLInputType
+        var type: GraphQLInputType? = defaultGraphQLScalars(property.returnType) as? GraphQLInputType
+        if (type == null) {
+            type = config.hooks.willGenerateGraphQLType(property.returnType)as? GraphQLInputType
```

Comment:
By proving the schemaconfig we can allow the user to use custom types in their custom directives via `Coercing` (e.g. `Permission` enums).